README
WWW-Scraper-ISBN-Pearson_Driver
===============================

This module is a Perl extension for ISBN search collection that is provide by
WWW::Scraper::ISBN.

This driver searches for book information from the Pearson Education online 
catalog.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules:

  WWW::Scraper::ISBN
  WWW::Scraper::ISBN::Record
  WWW::Scraper::ISBN::Driver
  WWW::Mechanize

For optional testing the module would like these other modules:

  Test::Pod
  Test::Pod::Coverage
  Test::CPAN::Meta
  Test::CPAN::Meta::JSON

COPYRIGHT AND LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.


Changes
Revision history for WWW-Scraper-ISBN-Pearson_Driver
====================================================

        - updated regex for site changes.

0.22    2014-11-30
        - fixed license fields in META.json to be lists.
        - ping fix for Cygwin (thanks to Alexandr Ciornii).

0.21    2014-07-29
        - fixed distribution name in META.
        - upgrade to WWW-Scraper-ISBN-1.00.
        - provide html in book details.

0.20    2013-11-27
        - META updates.
        - added git repository links to metadata.
        - added isbn-check.pl to examples.
        - added Data::Dumper to prerequisites.

0.19    2012-12-16
        - updated regex for site changes.
        - change file dates changed to meet W3CDTF standards.
        - updated form name for search.
        - License updated consistently to Artistic License v2.0.
        - reworked Makefile.PL for clarity.
        - META tests no longer require specific versions.
        - added minimum perl version (5.006).

0.18    2011-05-04
        - amended pingtest() in network tests.
        - added further skip test if test books are withdrawn from the site.
        - upgraded to Test::CPAN::Meta::JSON for tests.
        - image link updates due to site changes.

0.17    2010-10-06
        - provided alternate ping test methods.

0.16    2010-09-15
        - reworked test structure, so easier to add sample sets.

0.15    2010-08-10
        - added better protection from network failures.
        - updated prerequisite requirements.

0.14    2010-08-02
        - added connection protection if client or server offline.
        - documentation updates.

0.13    2010-07-30
        - documentation fixes.

0.12    2010-07-30
        - updates to parser for Pearson site.
        - added fields (if known) for binding, pages, weight, width & height.
        - added ean13, which is the industry name for isbn-13.
        - more tests.
        - removal of Template::Extract references.
        - various package file changes (inc adding META.json).

0.11	2009-08-14
        - minor fixes to parsing of Pearson website.
        - added ping tests, to avoid failing when no network connection.
        - real failures produce diagnostics not skips now.
        - added more package tests.

0.10	2008-06-27
        - added checks to only run extra tests under AUTOMATED_TESTING
        - switched metatest to use Test-CPAN-Meta
        - upped META.ym spec version and link
        - added isbn10 key into book data
        - attempt to catch bad ISBN in tests where Business::ISBN is installed,
          which is now used by the new version of WWW::Scraper::ISBN.

0.09	2007-07-26
        - added test to check for non-existent ISBNs
        - updated first book test to a new book, Perl Medic looks to be out of 
          print :(

0.08	2007-03-01
        - Handwritten META.yml to conform to current specs.
        - Changed all instances of LICENCE to LICENSE.
        - Added META.yml test script.
        - Added LICENSE and Artistic files.
        - Added an examples directory. Nothing in it yet, except a request
          for interesting scripts.
        - Updated the pattern matching code now that Yahoo have changed their 
          page layouts.
        - Added better error reporting.
        - Removed dependency on Template::Extract.
        - Devel::Cover = 95.2%

0.07	2005-03-02
        - book_link made from bookid stored in the page

0.06	2001-01-07  
        - handler() moved to WWW::Scraper::ISBN::Driver
        - test suite updated to peform skips if can't access web pages

0.05	2004-08-31
        - Simplified error handling

0.04	2004-06-07
        - Simplified extract string

0.03	2004-05-10
        - Added publisher attribute

0.02	2004-04-07
        - can't remember!

0.01	2004-04-07
        - original version


lib/WWW/Scraper/ISBN/Pearson_Driver.pm
package WWW::Scraper::ISBN::Pearson_Driver;

use strict;
use warnings;

use vars qw($VERSION @ISA);
$VERSION = '0.22';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::Pearson_Driver - Search driver for the Pearson Education online book catalog.

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from the Pearson Education's online catalog.

=cut

#--------------------------------------------------------------------------

###########################################################################
# Inheritence

use base qw(WWW::Scraper::ISBN::Driver);

###########################################################################
# Modules

use WWW::Mechanize;

###########################################################################
# Constants

use constant	SEARCH	=> 'http://www.pearsoned.co.uk/Bookshop/';
use constant	DETAIL	=> 'http://www.pearsoned.co.uk/Bookshop/detail.asp?item=';

#--------------------------------------------------------------------------

###########################################################################
# Public Interface

=head1 METHODS

=over 4

=item C<search()>

Creates a query string, then passes the appropriate form fields to the Pearson
Education server.

The returned page should be the correct catalog page for that ISBN. If not the
function returns zero and allows the next driver in the chain to have a go. If
a valid page is returned, the following fields are returned via the book hash:

  isbn          (now returns isbn13)
  isbn10        
  isbn13
  ean13         (industry name)
  author
  title
  book_link
  image_link
  description
  pubdate
  publisher
  binding       (if known)
  pages         (if known)
  weight        (if known) (in grammes)
  width         (if known) (in millimetres)
  height        (if known) (in millimetres)

The book_link and image_link refer back to the Pearson Education UK website.

=back

=cut

sub search {
	my $self = shift;
	my $isbn = shift;
	$self->found(0);
	$self->book(undef);

	my $mech = WWW::Mechanize->new();
    $mech->agent_alias( 'Linux Mozilla' );

	eval { $mech->get( SEARCH ) };
    return $self->handler("Pearson Education website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

#print STDERR "\n# content=[\n".$mech->content()."\n]\n";

	$mech->form_id('hipGlobalSearchForm');
	$mech->set_fields( 'txtSearch' => $isbn );
	
    eval { $mech->submit() };
	return $self->handler("Failed to find that book on Pearson Education website.")
	    if($@ || !$mech->success() || !$mech->content());

	# The Book page
    my $html = $mech->content();
#print STDERR "\n# html=[\n$html\n]\n";

	return $self->handler("Failed to find that book on Pearson Education website.")
		if($html =~ m!<p>Your search for <b>\d+</b> returned 0 results. Please search again.</p>!si);

    my $data;
    ($data->{image})        = $html =~ m!"(http://images.pearsoned-ema.com/jpeg/large/\d+\.jpg)"!i;
    ($data->{thumb})        = $html =~ m!"(http://images.pearsoned-ema.com/jpeg/small/\d+\.jpg)"!i;
    ($data->{title},
     $data->{author},
     $data->{pubdate},
     $data->{binding},
     $data->{pages},
     $data->{isbn13})       = $html =~ m!   <div\s*class="biblio">\s*
                                                        <h1\s*class="larger\s*bold">(.*?)</h1>\s*(?:.*?<br\s*/>\s*)?
                                                        <h2\s*class="body"><a\s*title[^>]+>(.*?)</a>\s*</h2>\s*
                                                        ([^,]+),\s*([^,]+)(?:,\s*(\d+)\s+pages)?(?:</a>)?<br\s*/>\s*
                                                        ISBN(?:13)?:\s*(\d+)\s*<br\s*/>!ix;
    ($data->{description})  = $html =~ m!<div class="desc-text"><p><p>([^<]+)!is;
    ($data->{bookid})       = $html =~ m!recommend.asp\?item=(\d+)!i;

#use Data::Dumper;
#print STDERR "\n# " . Dumper($data);

	return $self->handler("Could not extract data from Pearson Education result page.")
		unless(defined $data);

    # remove HTML tags
    for(qw(author binding)) {
        next unless(defined $data->{$_});
        $data->{$_} =~ s!<[^>]+>!!g;
    }

	# trim top and tail
	for(keys %$data) { 
        next unless(defined $data->{$_});
        $data->{$_} =~ s/^\s+//;
        $data->{$_} =~ s/\s+$//; 
    }

    my $uri = $mech->uri(); #DETAIL . $data->{bookid};

	my $bk = {
		'ean13'		    => $data->{isbn13},
		'isbn13'		=> $data->{isbn13},
		'isbn10'		=> $self->convert_to_isbn10( $data->{isbn13} ),
		'isbn'			=> $data->{isbn13},
		'author'		=> $data->{author},
		'title'			=> $data->{title},
		'book_link'		=> "$uri",
		'image_link'	=> $data->{image},
		'thumb_link'	=> $data->{thumb},
		'description'	=> $data->{description},
		'pubdate'		=> $data->{pubdate},
		'publisher'		=> q!Pearson Education!,
		'binding'	    => $data->{binding},
		'pages'		    => $data->{pages},
		'weight'		=> $data->{weight},
		'width'		    => $data->{width},
		'height'		=> $data->{height},
        'html'          => $html
	};

#use Data::Dumper;
#print STDERR "\n# book=".Dumper($bk);

    $self->book($bk);
	$self->found(1);
	return $self->book;
}

1;
__END__

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-Pearson_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  Miss Barbell Productions, <http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2004-2015 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut
